pkg/auth/csp/tanzu.go

```diff
@@ -0,0 +1,264 @@
+	clientID              string
+	listenAddr            string
+	callbackPath          string
+	oauthConfig           *oauth2.Config
+	pkceCodePair          pkce.Code
+	state                 state.State
+	token                 *oauth2.Token
+	refreshToken          string
+	orgID                 string
+}
+
+// LoginOption is an optional configuration for Login().
+type LoginOption func(*cspLoginHandler) error
+
+// WithRefreshToken causes the login to use refresh token instead of interactive login.
+// If the refresh token is expired or invalid, the interactive login will kick in
+func WithRefreshToken(refreshToken string) LoginOption {
+	return func(h *cspLoginHandler) error {
+		h.refreshToken = refreshToken
+		return nil
+	}
+}
+
+// WithOrgID causes the login to given Organization.
+func WithOrgID(orgID string) LoginOption {
+	return func(h *cspLoginHandler) error {
+		h.orgID = orgID
+		return nil
+	}
+}
+
+func (h *cspLoginHandler) handleTokenRefresh() (*Token, error) {
+	refreshedToken, err := h.oauthConfig.TokenSource(context.TODO(), &oauth2.Token{RefreshToken: h.refreshToken}).Token()
+	if err != nil {
+		return nil, err
+	}
+	return &Token{
+		IDToken:      refreshedToken.Extra("id_token").(string),
+		AccessToken:  refreshedToken.AccessToken,
+		RefreshToken: refreshedToken.RefreshToken,
+		ExpiresIn:    int64(time.Until(refreshedToken.Expiry).Seconds()),
+		TokenType:    IDTokenType,
+	}, nil
+}
+
+func TanzuLogin(issuerURL string, opts ...LoginOption) (*Token, error) {
+	h := &cspLoginHandler{
+		issuer:       issuerURL,
+		clientID:     tanzuCLIClientID,
+		listenAddr:   defaultListenAddress,
+		callbackPath: defaultCallbackPath,
+	}
+	h.oauthConfig = &oauth2.Config{
+		RedirectURL: (&url.URL{Scheme: "http", Host: h.listenAddr, Path: h.callbackPath}).String(),
+		ClientID:    h.clientID,
+		Endpoint: oauth2.Endpoint{
+			AuthURL:  KnownIssuers[issuerURL].AuthURL,
+			TokenURL: KnownIssuers[issuerURL].TokenURL,
+		},
+	}
+	for _, opt := range opts {
+		if err := opt(h); err != nil {
+			return nil, err
+		}
+	}
+
+	if h.refreshToken != "" {
+		// handle token refresh
+		token, err := h.handleTokenRefresh()
+		if err == nil {
+			return token, nil
+		}
+		// If refresh token fails, proceed with login flow through the browser
+	}
+
+	return h.handleBrowserLogin()
+}
+
+func (h *cspLoginHandler) handleBrowserLogin() (*Token, error) {
+	var err error
+	if h.pkceCodePair, err = pkce.Generate(); err != nil {
+		return nil, errors.Wrapf(err, "failed to generate pkce code pair generator")
+	}
+	h.state, err = state.Generate()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to generate state parameter")
+	}
+
+	shutdown, err := h.runLocalListener()
+	if err != nil {
+		return nil, err
+	}
+	defer shutdown()
+
+	authCodeURL := h.getAuthCodeURL()
+	log.Info("Opening the browser window to complete the login\n")
+	err = browser.OpenURL(authCodeURL)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open the browser for login")
```

Comment:
Not sure if you plan to handle this situation as a followup, but as it would be good to show the URL and urge the user to browse to it on his own. I suggest adding the TODO comment to summarize the plan here

Reply:
Sure. Added the TODO comment ( terminal based login usecase would be handled in the follow-up PR).